File: Second Seminar/SecondSeminar/src/main/java/com/example/SecondSeminar/common/exception/GlobalExceptionHandler.java
package com.example.SecondSeminar.common.exception;

import static org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {
    private static final String LOG_FORMAT = "%s";
    private static final String SERVER_ERROR_MESSAGE = "서버 오류가 발생하였습니다.";

    @ExceptionHandler(BaseCustomException.class)
    public ResponseEntity<ExceptionResponse> handleBaseException(BaseCustomException e) {
        ExceptionType exceptionType = e.getExceptionType();
        log.info(String.format(LOG_FORMAT, e.getMessage()), e);
        return ResponseEntity.status(exceptionType.httpStatus())
                .body(ExceptionResponse.of(exceptionType.name(), exceptionType.message()));
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ExceptionResponse> handleException(Exception e) {
        HttpStatus httpStatus = INTERNAL_SERVER_ERROR;
        log.error(String.format(LOG_FORMAT, e.getMessage()), e);
        return ResponseEntity.status(httpStatus)
                .body(ExceptionResponse.of(httpStatus.name(), SERVER_ERROR_MESSAGE));
    }
}

File: Second Seminar/SecondSeminarKotlin/src/main/kotlin/com/example/SecondSeminarKotlin/exception/GlobalExceptionHandler.kt
package com.example.SecondSeminarKotlin.exception

import org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

private const val SERVER_ERROR_MESSAGE = "서버 오류가 발생하였습니다"

@RestControllerAdvice
class GlobalExceptionHandler {
    @ExceptionHandler(BaseCustomException::class)
    protected fun handleBaseException(e: BaseCustomException): ResponseEntity<ExceptionResponse> {
        return ResponseEntity.status(e.getHttpStatus()).body(e.toErrorResponse())
    }

    @ResponseStatus(INTERNAL_SERVER_ERROR)
    @ExceptionHandler(RuntimeException::class)
    protected fun handleInternalServerError(e: RuntimeException): ResponseEntity<ExceptionResponse> {
        val httpStatus = INTERNAL_SERVER_ERROR
        return ResponseEntity.status(httpStatus)
            .body(ExceptionResponse.of(httpStatus.name, SERVER_ERROR_MESSAGE))
    }
}